File: app/src/main/java/val/mx/vbread/Complex.java
package val.mx.vbread;

public class Complex {
    private final Double real;
    private final Double imag;

    public Complex(Double real, Double imag) {
        this.real = real;
        this.imag = imag;
    }

    public Double getImag() {
        return imag;
    }

    public Double getReal() {
        return real;
    }

    public Double abs() {

        // WURZELBERECHNUNG https://de.wikipedia.org/wiki/Newtonverfahren
        // TODO: 02.11.2020

        double decimal = Math.pow(real, 2) + Math.pow(imag, 2);
        return Math.sqrt(decimal);
    }

    public Complex conjugate() {
        return new Complex(real,-imag);
    }

    public Complex multiply(Complex complex) {
        return new Complex(
                real * complex.real - imag * complex.imag,
                imag * complex.real + real * complex.imag);
    }

    public Complex multiply(double fac) {
        return new Complex(real * fac, imag * fac);
    }

    public Complex divide(Complex div) {
        System.out.println("1: " + div.conjugate());
        System.out.println("2: " + (Math.pow(div.imag,2)+Math.pow(div.real,2)));
        return multiply(div.conjugate()).divide(Math.pow(div.imag,2)+Math.pow(div.real,2));
    }

    public Complex divide(double fac) {

        return new Complex(real/fac,imag/fac);
    }

    public Complex pow(int fac) {
        Complex temp = this;
        for (int i = 0; i < fac; i++) {
            temp = temp.multiply(this);
        }
        return temp;
    }

    public Complex add(Double real) {
        return new Complex(this.real + real, imag);
    }

    public Complex add(Complex toAdd) {
        return new Complex(real + toAdd.real, imag + toAdd.imag);
    }

    public Complex subtract(Double real) {
        return new Complex(this.real - real, imag);
    }

    public Complex subtract(Complex toAdd) {
        return new Complex(real - toAdd.real, imag - toAdd.imag);
    }

    private Complex kehrwert() {
        double nenner = real*real + imag*imag;
        return new Complex(real/nenner,-imag/nenner);
    }

    // AB HIER "Kopiert" https://www.math.ksu.edu/~bennett/jomacg/c.html
    // TODO: 22.11.2020 Verstehen

    // Real cosh function (used to compute complex trig functions)

    private double cosh(double theta) {
        return (Math.exp(theta) + Math.exp(-theta)) / 2;
    }
    // Real sinh function (used to compute complex trig functions)

    private double sinh(double theta) {
        return (Math.exp(theta) - Math.exp(-theta)) / 2;
    }

    /**
     * Sine of this Complex number (doesn't change this Complex number).
     * <br>sin(z) = (exp(i*z)-exp(-i*z))/(2*i).
     *
     * @return sin(z) where z is this Complex number.
     */
    public Complex sin() {
        return new Complex(cosh(imag) * Math.sin(real), sinh(imag) * Math.cos(real));
    }

    /**
     * Cosine of this Complex number (doesn't change this Complex number).
     * <br>cos(z) = (exp(i*z)+exp(-i*z))/ 2.
     *
     * @return cos(z) where z is this Complex number.
     */
    public Complex cos() {
        return new Complex(cosh(imag) * Math.cos(real), -sinh(imag) * Math.sin(real));
    }

    @Override
    public String toString() {
        return "Complex{" +
                "real=" + real +
                ", imag=" + imag +
                '}';
    }

}

File: app/src/main/java/val/mx/vbread/Complex.kt
package `val`.mx.vbread

import kotlin.math.exp
import kotlin.math.pow
import kotlin.math.sqrt

class Complex(val real: Double, val imag: Double) {
    fun abs(): Double {
        val decimal = real.pow(2.0) + imag.pow(2.0)

//         double sqrt = Double.longBitsToDouble( ( ( Double.doubleToLongBits( decimal )-(1l<<52) )>>1 ) + ( 1l<<61 ) ); // WHAT https://stackoverflow.com/questions/13263948/fast-sqrt-in-java-at-the-expense-of-accuracy NOTE: NOT FAST
//         return sqrt;
        return sqrt(decimal)
    }

    private fun conjugate(): Complex {
        return Complex(real, -imag)
    }

    fun multiply(complex: Complex): Complex {
        return Complex(
            real * complex.real - imag * complex.imag,
            imag * complex.real + real * complex.imag
        )
    }

    fun multiply(fac: Double): Complex {
        return Complex(real * fac, imag * fac)
    }

    fun divide(div: Complex): Complex {
        return multiply(div.conjugate()).divide(div.imag.pow(2.0) + div.real.pow(2.0))
    }

    fun divide(fac: Double): Complex {
        return Complex(real / fac, imag / fac)
    }

    fun pow(fac: Int): Complex {
        var temp = this
        for (i in 0 until fac) {
            temp = temp.multiply(this)
        }
        return temp
    }

    fun add(real: Double): Complex {
        return Complex(this.real + real, imag)
    }

    fun add(toAdd: Complex): Complex {
        return Complex(real + toAdd.real, imag + toAdd.imag)
    }

    fun subtract(real: Double): Complex {
        return Complex(this.real - real, imag)
    }

    fun subtract(toAdd: Complex): Complex {
        return Complex(real - toAdd.real, imag - toAdd.imag)
    }

    private fun cosh(x: Double): Double {
        return (exp(x) + exp(-x)) / 2
    }

    private fun sinh(x: Double): Double {
        return (exp(x) - exp(-x)) / 2
    }

    fun sin(): Complex {
        return Complex(cosh(imag) * kotlin.math.sin(real), sinh(imag) * kotlin.math.cos(real))
    }

    fun cos(): Complex {
        return Complex(cosh(imag) * kotlin.math.cos(real), -sinh(imag) * kotlin.math.sin(real))
    }

    override fun toString(): String {
        return "Complex{" +
                "real=" + real +
                ", imag=" + imag +
                '}'
    }
}